integration_tests/testsuite/test_custom.py

```diff
@@ -39,19 +57,179 @@ def runTest(self):
+
+        logging.info('Running custom test: %s', name)
+
+        if self._test_for_old_specification(specification):
+            return
+
+        if validation is None:
+            self.fail('A validation must be specified in the step '
+                      'configuration')
+
+        context = {'name': name}
+        step_num = 0
+
+        for step in steps:
+            step_name, step_context = self._run_step(context, step,
+                                                     step_num, timeout)
+            context[step_name] = step_context
+
+        logging.debug('context : %s', json.dumps(context,
+                                                 sort_keys=True,
+                                                 indent=4,
+                                                 separators=(',', ': ')))
+
+        self._validate(context, validation)
+
+    def _run_step(self, context, step, step_num, timeout):
+        """Use the provided step's configuration to send a request to the
+           specified path and store the result into the context.
+
+        Args:
+            context: A dictionary containing the context for the test.
+            step: A dictionary containing the configuration of the step,
+               this include:
+                 name (optional): name of the step.
+                 configuration (optional):
+                    method: 'GET' or 'POST'.
+                    headers: Dictionary containing the headers of the request.
+                    content: Payload attached to the request.
+                 path: Url of the request
+            step_num: Index of the step.
+
+        Returns:
+            None.
+        """
+        step_name = step.get('name', 'step_{0}'.format(step_num))
+        configuration = step.get('configuration', dict())
+        path = step.get('path')
+
+        logging.info('Running step {0} of test {1}'.format(
+            step_name,
+            context.get('name')
+        ))
+
+        test_endpoint = urlparse.urljoin(self._base_url, path)
+        response = requests.request(method=configuration.get('method', 'GET'),
+                                    url=test_endpoint,
+                                    headers=configuration.get('headers'),
+                                    data=configuration.get('content'),
+                                    timeout=timeout)
+
+        if 'application/json' in response.headers.get('Content-Type'):
+            content = response.json()
+        else:
+            content = response.text
+
+        step_context = {
+            'request': {
+                'configuration': configuration,
+                'path': path
+            },
+            'response': {
+                'headers': dict(response.headers),
+                'status': response.status_code,
+                'content': content
+            }
+        }
+
+        return step_name, step_context
+
+    def _validate(self, context, specification):
```

Comment:
I find it confusing to call this member `specification` - as the word `specification` was previously used to describe a test step. I'd recommend refactoring to `validation` maybe `validation_spec`. And even above, for the test step, instead of `specification`, I'd reconsider naming it back to `test_info`, `step_specification`, `step_spec`, `test_step`, or something similar.